### PR TITLE
Fix multi call of defenderVotes in the society module of api-derive

### LINKED ISSUE #5766: RPC-CORE error when calling `api.derive.society.members()`
* **I'm submitting a ...**

  - [x] Bug report
  - [ ] Feature request
  - [ ] Support request
  - [ ] Other

* **What is the current behavior and expected behavior?**

The Society pallet was update to [V2](https://github.com/paritytech/substrate/pull/11324) after the release of runtime v1.0. The api was update to reflect the changes. The regular query api seems to be working fine, but the derive api is throwing a weird RPC-CORE error for me when calling members(). It worked normally before the update and I don't seem any changes to how I should call the method.

```
RPC-CORE: subscribeStorage(keys?: Vec<StorageKey>): StorageChangeSet:: createType(Vec<StorageKey>):: Expected an array of 2 values as params to a Map query
```

Full error stack:

```
type.js:54 Uncaught (in promise) Error: createType(Vec<StorageKey>):: Expected an array of 2 values as params to a Map query
    at createTypeUnsafe (type.js:54:1)
    at TypeRegistry.createTypeUnsafe (registry.js:330:1)
    at RpcCore._formatParams (bundle.js:338:1)
    at Observable._subscribe (bundle.js:286:1)
    at Observable._trySubscribe (Observable.js:33:1)
    at Observable.js:27:1
    at errorContext (errorContext.js:23:1)
    at Observable.subscribe (Observable.js:23:1)
    at catchError.js:9:1
    at OperatorSubscriber.<anonymous> (lift.js:10:1)
    at Observable.js:27:1
    at errorContext (errorContext.js:23:1)
    at Observable.subscribe (Observable.js:23:1)
    at distinctUntilChanged.js:12:1
    at OperatorSubscriber.<anonymous> (lift.js:10:1)
    at Observable.js:27:1
    at errorContext (errorContext.js:23:1)
    at Observable.subscribe (Observable.js:23:1)
    at ConnectableObservable.connect (ConnectableObservable.js:43:1)
    at Observable._subscribe (refCountDelay.js:15:1)
    at Observable._trySubscribe (Observable.js:33:1)
    at Observable.js:27:1
    at errorContext (errorContext.js:23:1)
    at Observable.subscribe (Observable.js:23:1)
    at combineLatest.js:42:1
    at maybeSchedule (combineLatest.js:68:1)
    at _loop_1 (combineLatest.js:39:1)
    at combineLatest.js:59:1
    at maybeSchedule (combineLatest.js:68:1)
    at Observable._subscribe (combineLatest.js:33:1)
    at Observable._trySubscribe (Observable.js:33:1)
    at Observable.js:27:1
    at errorContext (errorContext.js:23:1)
    at Observable.subscribe (Observable.js:23:1)
    at map.js:6:1
    at OperatorSubscriber.<anonymous> (lift.js:10:1)
    at Observable.js:27:1
    at errorContext (errorContext.js:23:1)
    at Observable.subscribe (Observable.js:23:1)
    at Observable._subscribe (memo.js:14:1)
    at Observable._trySubscribe (Observable.js:33:1)
    at Observable.js:27:1
    at errorContext (errorContext.js:23:1)
    at Observable.subscribe (Observable.js:23:1)
    at catchError.js:9:1
    at OperatorSubscriber.<anonymous> (lift.js:10:1)
    at Observable.js:27:1
    at errorContext (errorContext.js:23:1)
    at Observable.subscribe (Observable.js:23:1)
    at distinctUntilChanged.js:12:1
```

The error seems to be happening inside the `combineLast` call [around here](https://github.com/polkadot-js/api/blob/03cdccc56626949deea187383462ffbc8cdd0a44/packages/api-derive/src/society/members.ts#L39)

* **What is the motivation for changing the behavior?**

Fix [KappaSigmaMu website](https://ksmsociety.io/) to reflect the new changes to the Society pallet

* **Please tell us about your environment:**

  - Version:
  - Environment:

    - [x] Node.js (v20.10.0)
    - [x] Browser (Brave)
    - [ ] Other (limited support for other environments)

  - Language:

    - [ ] JavaScript
    - [x] TypeScript (include tsc --version)
    - [ ] Other

### PR DESCRIPTION
**Pull request by laurogripa**
Closes #5766

### CODE REVIEW

File: packages/api-derive/src/society/members.ts [hunk 1]
```diff
@@ -36,11 +36,16 @@ function _membersPrev (api: DeriveApi, accountIds: AccountId[]): Observable<Deri
...
+  api.query.society.challengeRoundCount().subscribe(result => {
+    currentChallengeRound = result;
+  });
+
   return combineLatest([
     of(accountIds),
     api.query.society.members.multi(accountIds),
     api.query.society.payouts.multi(accountIds),
-    api.query.society.defenderVotes.multi(accountIds),
+    api.query.society.defenderVotes.multi(accountIds.map(accountId => [currentChallengeRound, accountId])),
```
Comment (jacogr) [inline packages/api-derive/src/society/members.ts:48]:
```suggestion
    api.query.society.challengeRoundCount().pipe(
      switchMap((round) =>
        api.query.society.defenderVotes.multi(accountIds.map(accountId => [round, accountId]))
      )
    ),
```

I believe this should address the issue better - the round is retrieved and then fed to the query. (And all return results at the same time). This gets around the issue where we do a query for round 0 first.

**Review by jacogr — CHANGES_REQUESTED** [review]

File: packages/api-derive/src/society/members.ts (hunk 1 above)
Comment (laurogripa) [reply packages/api-derive/src/society/members.ts]: Nice suggestion! Looks cleaner.

**Comment by laurogripa** [comment]
Thanks for reviewing @jacogr.

I wanted to add a regression test too, I even [committed a first version](https://github.com/polkadot-js/api/blob/647a8afc9ecf3df84ecb672e7dbdc6dc159b5efa/packages/api-derive/src/society/members.spec.ts), but I couldn't figure out how to mock the `society.members()` response from the RPC. Without a mock the members array is empty and the error on defenderVotes doesn't get triggered.

Let me know if there's an easy way to mock this response and I can add the test back.
If you don't mind the absence of test, you can go ahead and merge, I tested manually and seems to be working just fine.

**Comment by laurogripa** [comment]
By the way, I just found another smaller issue and committed a change...

I noticed that `isDefenderVoter` is only true if the member vote for the approval of the Defender.
In my opinion we should return true if the member vote, doesn't matter if it's an approval or a rejection, just like it was before with `_membersPrev`

**Comment by TarikGul** [comment]
> By the way, I just found another smaller issue and committed a change...
> 
> I noticed that `isDefenderVoter` is only true if the member voted for the approval of the Defender. In my opinion we should return true if the member voted, doesn't matter if it's an approval or a rejection, just like it was before with `_membersPrev`

I agree that the change here makes sense. Nice catch, I am not sure what the original purpose was around having the `isDefenderVoter` point to whether the vote was an approval for `_membersCurr`, but it doesn't seem to be in line with what `_membersPrev` is or the naming of the key so nice catch .

**Review by TarikGul — APPROVED** [review]
lgtm, I'll make sure to highlight the changes in the next release.

**Comment by TarikGul** [comment]
@laurogripa, due to some circumstances can you create a new PR with a description of the changes, and I'll give it a check and we can merge it in?

**Comment by laurogripa** [comment]
> @laurogripa, due to some circumstances can you create a new PR with a description of the changes, and I'll give it a check and we can merge it in?

Hey, Tarik, thanks for reviewing. Does it need to be a new PR, or just an update of the description is enough?

**Comment by TarikGul** [comment]
> > @laurogripa, due to some circumstances can you create a new PR with a description of the changes, and I'll give it a check and we can merge it in?
> 
> Hey, @TarikGul, thanks for reviewing. Does it need to be a new PR, or just an update of the description is enough?

A new PR would be amazing, I know it's not ideal but there is no way around the Blocked Merging currently.

**Comment by TarikGul** [comment]
@laurogripa Still available to port this to a new PR?

**Comment by laurogripa** [comment]
Yes, sorry for the delay, haven’t had a lot of time lately, but I’ll
arrange to port to a new PR in the next few days

**Pull request closed without merging**